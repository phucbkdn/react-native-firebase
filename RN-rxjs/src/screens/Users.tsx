import React, { FC, useState, useEffect } from 'react'
import { View, Text, FlatList, TouchableOpacity, Alert } from 'react-native'
import { useNavigation } from '@react-navigation/native'
import firebaseApp from '../services'
import { messagesStyles } from './styles/Messages.styles'
import { Screen } from '../components/screen/screen'
import { color } from '../themes'
import { User } from '../components/User'
import usersStore from '../store/users'
import { User as UserType } from '../models'

const Users: FC = () => {
  const [users, setUsers] = useState<UserType[]>([])
  const navigation = useNavigation()

  useEffect(() => {
    const sub = usersStore.getStore().subscribe((it) => {
      setUsers(it.users)
    })
    usersStore.init()

    return () => sub.unsubscribe()
  }, [])

  const handleLogout = () => {
    firebaseApp
      .auth()
      .signOut()
      .then(() => {
        console.log('Sign-out successful.')
        navigation.navigate('Root')
      })
      .catch((err) => {
        Alert.alert('Error', err.message)
      })
  }

  return (
    <View style={messagesStyles.container}>
      <Screen
        style={messagesStyles.screen}
        preset="fixed"
        backgroundColor={color.transparent}
      >
        <Text style={messagesStyles.title}>Chat App</Text>
        <TouchableOpacity onPress={handleLogout}>
          <Text>Logout</Text>
        </TouchableOpacity>
        <FlatList
          data={users}
          renderItem={({ item }) => (
            <User
              active={item.state === 'online'}
              id={item.id}
              name={item.email}
            />
          )}
          keyExtractor={(item) => item.id}
        />
      </Screen>
    </View>
  )
}

export default Users
