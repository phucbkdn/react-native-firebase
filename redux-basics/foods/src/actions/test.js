var obj = {a : 0, b : 2};
var copy =Object.assign({}, obj, { a: 1});
console.log(copy);