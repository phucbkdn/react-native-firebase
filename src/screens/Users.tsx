import React, { FC, useState, useEffect } from 'react'
import {
  View,
  Text,
  FlatList,
  TouchableOpacity
} from 'react-native'
import { list } from 'rxfire/database'
import { useNavigation } from '@react-navigation/native'
import { combineLatest } from 'rxjs'
import { map, filter } from 'rxjs/operators'
import { authState } from 'rxfire/auth'
import firebaseApp from '../services'
import { messagesStyles } from './styles/Messages.styles'
import { Screen } from '../components/screen/screen'
import { color } from '../themes'
import { User } from '../components/User'

const Users: FC = () => {
  const [users, setUsers] = useState([])
  const navigation = useNavigation()

  useEffect(() => {
    const ref = firebaseApp.database().ref('users')
    const auth = firebaseApp.auth()

    const user$ = authState(auth).pipe(filter(user => !!user))
    const users$ = list(ref)
    .pipe(
      map(changes => changes.map(c => {
        return { id: c.snapshot.key, ...c.snapshot.val() }
      }))
    )
    const subject = combineLatest(users$, user$)
    subject.subscribe(([items, item]) => {
      // Remove current user
      const result = items.filter(t => t.email !== item.email)
      setUsers(result)
    })
  }, [])

  const handleLogout = () => {
    firebaseApp.auth().signOut().then(() => {
      console.log('Sign-out successful.')
      navigation.navigate('Root')
    }).catch(err => {
      console.log(err)
    })
  }

  return (
    <View style={messagesStyles.container}>
      <Screen
        style={messagesStyles.screen}
        preset="fixed"
        backgroundColor={color.transparent}
      >
        <Text style={messagesStyles.title}>Chat App</Text>
        <TouchableOpacity onPress={handleLogout}>
          <Text>Logout</Text>
        </TouchableOpacity>
        <FlatList
          data={users}
          renderItem={({ item }) =>
          (<User active={item.state === 'online'} id={item.id} name={item.email} />)}
          keyExtractor={item => item.id}
        />
      </Screen>
    </View>
  )
}

export default Users
