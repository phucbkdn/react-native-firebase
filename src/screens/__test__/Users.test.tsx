import React from 'react'
import renderer from 'react-test-renderer'
import MockedNavigator from '../../test/mock-navigator'
import Users from '../Users'
import { setupFirebaseUnitTest, teardown } from '../../test/setupFirebaseTest'

const { adminDB } = setupFirebaseUnitTest()

beforeEach(async () => {
  await adminDB.collection('users').add({
    state: 'online',
    email: 'abc1@test.com',
  })
  await adminDB.collection('users').add({
    state: 'online',
    email: 'test@test.com',
  })
})

afterEach(async () => {
  await teardown()
})

describe('Testing Users screen', () => {
  const Component = () => <Users />
  const wrapper = renderer.create(<MockedNavigator component={Component} />)
  test('Render correct component', async () => {
    expect(wrapper).toMatchSnapshot()
  })
})
