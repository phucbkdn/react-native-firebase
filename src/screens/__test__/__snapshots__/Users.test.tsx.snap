// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Users screen Render correct component 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      pointerEvents="box-none"
      style={
        Array [
          Object {
            "zIndex": 1,
          },
          false,
        ]
      }
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        onLayout={[Function]}
        pointerEvents="box-none"
        style={null}
      >
        <View
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": 0,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgb(255, 255, 255)",
                "borderBottomColor": "rgb(216, 216, 216)",
                "flex": 1,
                "shadowColor": "rgb(216, 216, 216)",
                "shadowOffset": Object {
                  "height": 0.5,
                  "width": 0,
                },
                "shadowOpacity": 0.85,
                "shadowRadius": 0,
              }
            }
          />
        </View>
        <View
          pointerEvents="box-none"
          style={
            Object {
              "height": 64,
              "maxHeight": undefined,
              "minHeight": undefined,
              "opacity": undefined,
              "transform": undefined,
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "height": 20,
              }
            }
          />
          <View
            pointerEvents="box-none"
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "marginHorizontal": 16,
                  "opacity": 1,
                }
              }
            >
              <Text
                accessibilityRole="header"
                aria-level="1"
                numberOfLines={1}
                onLayout={[Function]}
                style={
                  Object {
                    "color": "rgb(28, 28, 30)",
                    "fontSize": 17,
                    "fontWeight": "600",
                  }
                }
              >
                Root
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "opacity": 1,
            }
          }
        />
        <View
          accessibilityElementsHidden={false}
          closing={false}
          gestureVelocityImpact={0.3}
          importantForAccessibility="auto"
          onClose={[Function]}
          onGestureBegin={[Function]}
          onGestureCanceled={[Function]}
          onGestureEnd={[Function]}
          onOpen={[Function]}
          onTransitionStart={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          transitionSpec={
            Object {
              "close": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
              "open": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 10,
                  "restSpeedThreshold": 10,
                  "stiffness": 1000,
                },
              },
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              collapsable={false}
              forwardedRef={[Function]}
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              style={
                Object {
                  "flex": 1,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <View
                pointerEvents="none"
                style={
                  Object {
                    "backgroundColor": "rgb(242, 242, 242)",
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": -1,
                    },
                    "shadowOpacity": 0.3,
                    "shadowRadius": 5,
                    "top": 0,
                    "width": 3,
                  }
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "overflow": "hidden",
                    },
                    Array [
                      Object {
                        "backgroundColor": "rgb(242, 242, 242)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "flexDirection": "column-reverse",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        onLayout={[Function]}
                        style={
                          Array [
                            Array [
                              Object {
                                "backgroundColor": "#ffffff",
                                "flex": 1,
                                "height": "100%",
                              },
                              Object {
                                "backgroundColor": "rgba(0, 0, 0, 0)",
                              },
                            ],
                            Object {
                              "paddingBottom": 0,
                            },
                          ]
                        }
                      >
                        <View
                          style={
                            Array [
                              Object {
                                "alignItems": "stretch",
                                "height": "100%",
                                "justifyContent": "flex-start",
                                "width": "100%",
                              },
                              Object {
                                "flex": 1,
                              },
                              Object {
                                "paddingTop": 20,
                              },
                            ]
                          }
                        >
                          <View
                            style={
                              Object {
                                "alignItems": "center",
                                "flexDirection": "row",
                                "paddingBottom": 8,
                                "paddingLeft": 15,
                                "paddingRight": 15,
                                "paddingTop": 6,
                              }
                            }
                          >
                            <View
                              style={
                                Object {
                                  "marginRight": 12,
                                }
                              }
                            >
                              <View
                                style={
                                  Object {
                                    "backgroundColor": "transparent",
                                    "borderRadius": 37.5,
                                    "height": 75,
                                    "width": 75,
                                  }
                                }
                                theme={
                                  Object {
                                    "colors": Object {
                                      "black": "#242424",
                                      "disabled": "hsl(208, 8%, 90%)",
                                      "divider": "#bcbbc1",
                                      "error": "#ff190c",
                                      "grey0": "#393e42",
                                      "grey1": "#43484d",
                                      "grey2": "#5e6977",
                                      "grey3": "#86939e",
                                      "grey4": "#bdc6cf",
                                      "grey5": "#e1e8ee",
                                      "greyOutline": "#bbb",
                                      "platform": Object {
                                        "android": Object {
                                          "error": "#f44336",
                                          "grey": "rgba(0, 0, 0, 0.54)",
                                          "primary": "#2196f3",
                                          "secondary": "#9C27B0",
                                          "success": "#4caf50",
                                          "warning": "#ffeb3b",
                                        },
                                        "ios": Object {
                                          "error": "#ff3b30",
                                          "grey": "#7d7d7d",
                                          "primary": "#007aff",
                                          "searchBg": "#dcdce1",
                                          "secondary": "#5856d6",
                                          "success": "#4cd964",
                                          "warning": "#ffcc00",
                                        },
                                      },
                                      "primary": "#2089dc",
                                      "searchBg": "#303337",
                                      "secondary": "#ca71eb",
                                      "success": "#52c41a",
                                      "warning": "#faad14",
                                      "white": "#ffffff",
                                    },
                                  }
                                }
                              >
                                <View
                                  accessibilityIgnoresInvertColors={true}
                                  style={
                                    Object {
                                      "backgroundColor": "transparent",
                                      "borderRadius": 37.5,
                                      "flex": 1,
                                      "overflow": "hidden",
                                      "position": "relative",
                                    }
                                  }
                                >
                                  <Image
                                    borderRadius={37.5}
                                    onLoad={[Function]}
                                    source={
                                      Object {
                                        "uri": "https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_1280.png",
                                      }
                                    }
                                    style={
                                      Object {
                                        "bottom": 0,
                                        "flex": 1,
                                        "height": null,
                                        "left": 0,
                                        "position": "absolute",
                                        "right": 0,
                                        "top": 0,
                                        "width": null,
                                      }
                                    }
                                    testID="RNE__Image"
                                    theme={
                                      Object {
                                        "colors": Object {
                                          "black": "#242424",
                                          "disabled": "hsl(208, 8%, 90%)",
                                          "divider": "#bcbbc1",
                                          "error": "#ff190c",
                                          "grey0": "#393e42",
                                          "grey1": "#43484d",
                                          "grey2": "#5e6977",
                                          "grey3": "#86939e",
                                          "grey4": "#bdc6cf",
                                          "grey5": "#e1e8ee",
                                          "greyOutline": "#bbb",
                                          "platform": Object {
                                            "android": Object {
                                              "error": "#f44336",
                                              "grey": "rgba(0, 0, 0, 0.54)",
                                              "primary": "#2196f3",
                                              "secondary": "#9C27B0",
                                              "success": "#4caf50",
                                              "warning": "#ffeb3b",
                                            },
                                            "ios": Object {
                                              "error": "#ff3b30",
                                              "grey": "#7d7d7d",
                                              "primary": "#007aff",
                                              "searchBg": "#dcdce1",
                                              "secondary": "#5856d6",
                                              "success": "#4cd964",
                                              "warning": "#ffcc00",
                                            },
                                          },
                                          "primary": "#2089dc",
                                          "searchBg": "#303337",
                                          "secondary": "#ca71eb",
                                          "success": "#52c41a",
                                          "warning": "#faad14",
                                          "white": "#ffffff",
                                        },
                                      }
                                    }
                                    transition={true}
                                  />
                                  <View
                                    accessibilityElementsHidden={true}
                                    importantForAccessibility="no-hide-descendants"
                                    pointerEvents="none"
                                    style={
                                      Object {
                                        "bottom": 0,
                                        "left": 0,
                                        "opacity": 1,
                                        "position": "absolute",
                                        "right": 0,
                                        "top": 0,
                                      }
                                    }
                                  >
                                    <View
                                      style={
                                        Object {
                                          "alignItems": "center",
                                          "backgroundColor": "#bdbdbd",
                                          "flex": 1,
                                          "height": null,
                                          "justifyContent": "center",
                                          "width": null,
                                        }
                                      }
                                      testID="RNE__Image__placeholder"
                                    />
                                  </View>
                                  <View
                                    style={
                                      Object {
                                        "flex": 1,
                                        "height": null,
                                        "width": null,
                                      }
                                    }
                                  />
                                </View>
                              </View>
                            </View>
                            <View>
                              <Text
                                style={
                                  Object {
                                    "fontSize": 16,
                                  }
                                }
                              />
                              <Text
                                style={
                                  Object {
                                    "color": "gray",
                                    "fontSize": 16,
                                  }
                                }
                              />
                            </View>
                          </View>
                          <RCTScrollView
                            data={
                              Array [
                                Object {
                                  "email": "abc1@test.com",
                                  "id": "9NURUowzheYhFOyWSwJ7",
                                  "state": "online",
                                },
                              ]
                            }
                            disableVirtualization={false}
                            getItem={[Function]}
                            getItemCount={[Function]}
                            horizontal={false}
                            initialNumToRender={10}
                            keyExtractor={[Function]}
                            maxToRenderPerBatch={10}
                            onContentSizeChange={[Function]}
                            onEndReachedThreshold={2}
                            onLayout={[Function]}
                            onMomentumScrollEnd={[Function]}
                            onScroll={[Function]}
                            onScrollBeginDrag={[Function]}
                            onScrollEndDrag={[Function]}
                            removeClippedSubviews={false}
                            renderItem={[Function]}
                            scrollEventThrottle={50}
                            stickyHeaderIndices={Array []}
                            updateCellsBatchingPeriod={50}
                            viewabilityConfigCallbackPairs={Array []}
                            windowSize={21}
                          >
                            <View>
                              <View
                                onLayout={[Function]}
                                style={null}
                              >
                                <View>
                                  <View
                                    accessible={true}
                                    focusable={true}
                                    onClick={[Function]}
                                    onResponderGrant={[Function]}
                                    onResponderMove={[Function]}
                                    onResponderRelease={[Function]}
                                    onResponderTerminate={[Function]}
                                    onResponderTerminationRequest={[Function]}
                                    onStartShouldSetResponder={[Function]}
                                    style={
                                      Object {
                                        "alignItems": "center",
                                        "flex": 1,
                                        "flexDirection": "row",
                                        "margin": 10,
                                        "opacity": 1,
                                        "paddingBottom": 20,
                                        "width": 750,
                                      }
                                    }
                                  >
                                    <Text
                                      allowFontScaling={false}
                                      style={
                                        Array [
                                          Object {
                                            "color": undefined,
                                            "fontSize": 30,
                                          },
                                          Object {
                                            "color": "green",
                                            "marginBottom": -3,
                                          },
                                          Object {
                                            "fontFamily": "FontAwesome",
                                            "fontStyle": "normal",
                                            "fontWeight": "normal",
                                          },
                                          Object {},
                                        ]
                                      }
                                    >
                                      
                                    </Text>
                                    <Text
                                      style={
                                        Object {
                                          "color": "#999",
                                          "fontSize": 11,
                                          "lineHeight": 13,
                                          "paddingLeft": 15,
                                          "textTransform": "capitalize",
                                          "width": 562.5,
                                        }
                                      }
                                    >
                                      abc1@test.com
                                    </Text>
                                  </View>
                                </View>
                              </View>
                            </View>
                          </RCTScrollView>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
